python/pyarrow/pandas_compat.py

```diff
@@ -365,6 +364,44 @@ def dataframe_to_arrays(df, schema, preserve_index, nthreads=1, columns=None,
         name = _index_level_name(column, i, column_names)
         index_column_names.append(name)
 
+    names = column_names + index_column_names
+
+    return (names, column_names, index_columns, index_column_names,
+            columns_to_convert, convert_types)
+
+
+def dataframe_to_types(df, preserve_index, columns=None):
+    names, column_names, index_columns, index_column_names, \
+        columns_to_convert, _ = _get_columns_to_convert(
+            df, None, preserve_index, columns
+        )
+
+    types = []
+    # If pandas knows type, skip conversion
+    for c in columns_to_convert:
+        values = c.values
+        if isinstance(c.values, pd.Categorical):
+            type_ = pa.array(c, from_pandas=True).type
+        else:
+            values, type_ = get_datetimetz_type(values, c.dtype, None)
+            type_ = pa.lib._ndarray_to_arrow_type(values, type_)
+            if type_ is None:
+                type_ = pa.array(c, from_pandas=True).type
+        types.append(type_)
```

Comment:
Well, this is certainly not cheap to compute in all cases. I think it's fine to focus on getting the API and behavior correct, then focusing later on performance

Reply:
Yeah, my intention is to finalize this PR and improve it later.